stdlib/Random/src/RNGs.jl

```diff
@@ -285,14 +285,66 @@ function seed!(r::MersenneTwister, seed::Vector{UInt32})
     return r
 end
 
-seed!(r::MersenneTwister=GLOBAL_RNG) = seed!(r, make_seed())
+seed!(r::MersenneTwister=get_local_rng()) = seed!(r, make_seed())
 seed!(r::MersenneTwister, n::Integer) = seed!(r, make_seed(n))
-seed!(seed::Union{Integer,Vector{UInt32}}) = seed!(GLOBAL_RNG, seed)
+seed!(seed::Union{Integer,Vector{UInt32}}) = seed!(get_local_rng(), seed)
 
 
-### Global RNG (must be defined after seed!)
+### Global RNG
 
-const GLOBAL_RNG = MersenneTwister(0)
+const THREAD_RNGs = MersenneTwister[]
+@inline get_local_rng() = get_local_rng(Threads.threadid())
+@noinline Base.@pure function get_local_rng(tid::Int)
+    #tls = task_local_storage()
+    #RNG = get(tls, :RNG, nothing)
+    #RNG isa MersenneTwister && return RNG
+    if length(THREAD_RNGs) < tid
+        resize!(THREAD_RNGs, Threads.nthreads())
```

Comment:
I'm not sure whether this is thread safe.

Reply:
Yeah, I still need to push my local fork which fixes some stuff like this. Was just waiting on the outcome of the naming discussion.